virttest/qemu_devices/qcontainer.py

```diff
@@ -1720,6 +1720,147 @@ def images_define_by_params(self, name, image_params, media=None,
+                    pci_bus = {'aobject': 'pci.0'}
+                bus = self.get_first_free_bus(
+                    {'type': 'SERIAL', 'atype': serial_type}, [None, nr])
+                if bus is None:
+                    _hba = serial_type.replace('-', '_') + '%s'
+                    bus = self.idx_of_next_named_bus(_hba)
+                    bus = self.list_missing_named_buses(_hba, 'SERIAL', bus + 1)[-1]
+                    logging.debug("list missing named bus: %s", bus)
+                    devices.append(
+                        qdevices.QDevice(serial_type,
+                                         {"id": bus},
+                                         bus,
+                                         pci_bus,
+                                         qdevices.QSerialBus(bus, serial_type, bus)))
+                else:
+                    bus = bus.busid
+            devices.append(
+                qdevices.QDevice(port_type,
+                                 {"id": serial_id},
+                                 parent_bus={'busid': bus}))
+            devices[-1].set_param('name', serial_name)
+        else:  # none virtio type, generate serial device directly
+            devices.append(qdevices.QDevice(serial_type, {"id": serial_id}))
+            devices[-1].set_param("reg", reg)
+        devices[-1].set_param('chardev', chardev_id)
+
+        return devices
+
+    def serials_define_by_params(self, serial_id, serial_params,
+                                 file_name=None, index=None):
+        """
+        Wrapper for creating serial device from serial params.
+
+        :param serial_id: id of serial object
+        :param serial_params: serial params
+        :param file_name: the file path of the serial device (optional)
+        :param index: the index of the serial device (optional)
+        :return: the device list that construct the serial device
+        """
+        serial_type = serial_params['serial_type']
+        virtio_port_type = serial_params.get('virtio_port_type')
+        machine = serial_params.get('machine_type')
+        if virtio_port_type:
+            if "-mmio" in machine:
+                controller_suffix = 'device'
+            elif machine.startswith("s390"):
+                controller_suffix = 'ccw'
+            else:
+                controller_suffix = 'pci'
+            serial_type = '%s-%s' % (serial_type, controller_suffix)
+            virtio_port_type = "virt%s" % virtio_port_type
+
+        prefix = serial_params.get('virtio_port_name_prefix')
+        name = serial_id
+        name = prefix + name if prefix else name
+        serial_params['name'] = name
+
+        chardev_args = {'logfile': serial_params.get('chardev_logfile'),
+                        'logappend': serial_params.get('chardev_logappend')}
+        backend = serial_params.get('chardev_backend', 'unix_socket')
+        if backend not in ['socket', 'vc', 'ringbuf',
+                           'spiceport', 'spicevmc']:
```

Comment:
`socket` and  `vc`  have option `mux`, why exclude them ?

Reply:
Default value for `backend` is `unix_socket`,  but here it use `['socket', 'vc', 'ringbuf','spiceport', 'spicevmc']`.

Reply:
@yanan-fu That's the condition statement for the option `mux`, nothing to do with what is default value.

Reply:
@yanan-fu Ah I get your point, `socket` should be `unix_socket`, but yes,  `socket` and `vc` have the option, should be removed from this list. I just copy it from CharDevice, not confirm all of them, my careless. Thank you for being so careful.